src/trace_ng/log.zig

```diff
@@ -0,0 +1,600 @@
+// TODO Improve import.
+const Channel = @import("../sync/channel.zig").Channel;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+const AtomicBool = std.atomic.Value(bool);
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+pub const Config = struct {
+    max_level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    /// Maximum memory that logger can use.
+    max_buffer: u64,
+    exit_sig: *std.atomic.Value(bool),
+    kind: LogKind = LogKind.standard,
+};
+
+const INITIAL_LOG_CHANNEL_SIZE: usize = 1024;
+
+const LogKind = enum {
+    standard,
+    testing,
+    noop,
+};
+
+const UnscopedLogger = StandardLogger(null);
+pub fn StandardLogger(comptime scope: ?type) type {
+    const StanardErrLogger = struct {
+        const Self = @This();
+        max_level: Level,
+        exit_sig: *std.atomic.Value(bool),
+        allocator: Allocator,
+        recycle_fba: RecycleFBA,
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) Self {
+            return .{
+                .allocator = config.allocator,
+                .recycle_fba = RecycleFBA.init(config.allocator, config.max_buffer) catch @panic("could not create RecycleFBA"),
+                .max_buffer = config.max_buffer,
+                .max_level = config.max_level,
+                .exit_sig = config.exit_sig,
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = null,
+            };
+        }
+
+        fn unscoped(self: *Self) *UnscopedLogger {
+            return @ptrCast(self);
+        }
+
+        fn withScope(self: *Self, comptime new_scope: anytype) *StandardLogger(new_scope) {
+            return @ptrCast(self);
+        }
+
+        pub fn spawn(self: *Self) void {
+            self.handle = std.Thread.spawn(.{}, Self.run, .{self}) catch @panic("could not spawn Logger");
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (self.handle) |*handle| {
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.close();
+            self.channel.deinit();
+            self.recycle_fba.deinit();
+        }
+
+        pub fn run(self: *Self) void {
+            while (!self.exit_sig.load(.seq_cst)) {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                const messages = self.channel.drain() orelse {
+                    // channel is closed
+                    return;
+                };
+                defer self.channel.allocator.free(messages);
+
+                for (messages) |message| {
+                    const writer = std.io.getStdErr().writer();
+                    logfmt.writeLog(writer, message) catch @panic("logging failed");
+                }
+                self.recycle_fba.mux.lock();
+                self.recycle_fba.alloc_allocator.reset();
```

Comment:
Not sure if this is the idiomatic way to do this, but without this it seems the space allocated is not reclaimed and given multiple allocation the recycle_fba get stuck when trying to allocate.

Reply:
This could corrupt any allocations that have not been freed yet, and any future allocations that use the same region of memory.